Add webauthn support for the two step re-auth dialog

xknown opened the pull request:
### Changes proposed in this Pull Request

* Adds support for security keys on the `2fa-reauth-dialog` block. 

### Testing instructions

* Add a security key to your WP.com account.
* Navigate to `/me` and remove the `twostep_auth` cookie from the browser console.
* Reload the page, and notice the `2fa-reauth-dialog` is shown along with a new option to authenticate using a hardware key. 

### Screenshots
#### Re-auth dialog
**Before**
<img  width="436" src="https://user-images.githubusercontent.com/1081870/74947345-b399ff00-53c8-11ea-99b7-a47073a4d6f9.png" />

**After**
<img  width="436" src="https://user-images.githubusercontent.com/1081870/74947405-cf050a00-53c8-11ea-8ed2-741c6a1cc437.png" />

when someone clicks the `Continue with security key` button

<img  width="436" src="https://user-images.githubusercontent.com/1081870/74947382-c14f8480-53c8-11ea-9e8d-a64f3df8fc84.png" />

xknown committed:
Add support for webauthn on the reauth-required block

xknown committed:
Implement the webauthn authentication flow.

xknown committed:
Instruct the backend to only generate a 2fa cookie.

xknown committed:
Do not do a remote request if we know it's going to fail.

xknown committed:
Enable the `Continue via text message` button even if it's rate limited

xknown committed:
Remove unused code

xknown committed:
Remove unneeded paragraph tag

tyxla reviewed (commented):
When testing this PR (in both calypso.localhost and calypso.live) I'm unable to set the twoauth step auth properly. I've tried multiple times to input my authenticator code, and then after a refresh, I get prompted to enter it again. Can't reproduce this behavior on `master`.

xknown commented:
> When testing this PR (in both calypso.localhost and calypso.live) I'm unable to set the twoauth step auth properly. I've tried multiple times to input my authenticator code, and then after a refresh, I get prompted to enter it again. Can't reproduce this behavior on `master`.

I just tested it on `https://hash-812da57f6b5a3de17b6d54ba6ef28cfc3eba53c2.calypso.live/me` and I'm not able to reproduce the problem. Can you please share how are you testing this? And, is there something mentioned in the browser console or the network requests?

Also, if you are testing this on safari, ITP protection might be the culprit.

tyxla commented on client/me/reauth-required/security-key-form.jsx:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Internal dependencies
+ */
+import { Card } from '@automattic/components';
```
Nit: for better readability and predictability of the code, we often sort the imports alphabetically.

xknown replied:
Fixed in 609c5af60131cb75c81efcc03735f7780cf898e1

tyxla commented on client/me/reauth-required/security-key-form.jsx:
```diff
@@ -0,0 +1,109 @@
+import FormButton from 'components/forms/form-button';
+import { localize } from 'i18n-calypso';
+import Spinner from 'components/spinner';
+import { recordTracksEventWithClientId as recordTracksEvent } from 'state/analytics/actions';
```
Since there is a `recordTracksEvent`, we should be careful with aliasing something else to that name. We could just reference this as `this.props. recordTracksEventWithClientId` in the component and that would do the job.

tyxla commented on client/me/reauth-required/security-key-form.jsx:
```diff
@@ -0,0 +1,109 @@
+	static propTypes = {
+		loginUserWithSecurityKey: PropTypes.func.isRequired,
+		onComplete: PropTypes.func,
+		recordTracksEvent: PropTypes.func.isRequired,
```
We generally differentiate between the three types of props:

* Expected props - props that are passed to the component when it's rendered somewhere.
* Connected props - props that are returned from the `connect` call, generally returned in `mapStateToProps` or `mapDispatchToProps`
* High-order component props - props that are being provided by high order components.

For example, we would often do this:

```
static propTypes = {
	expectedProp: PropTypes.string,

	// Connected props
	foo: PropTypes.object,
	bar: PropTypes.string,

	// From localize() HoC
	translate: PropTypes.func.isRequired,

	// From withLocalizedMoment() HoC
	moment: PropTypes.func.isRequired,
};
```

xknown replied:
Addressed this and the other comments below in c1184303d09f26a3ed772567df78b4251b6cdb7c

tyxla commented on client/me/reauth-required/security-key-form.scss:
```diff
@@ -0,0 +1,21 @@
+.security-key-form__verification-code-form .button {
```
Would it be better if we wrap those like this:

```
.security-key-form__verification-code-form {
	.button {
		...
	}
	p {
		...
	}
}
```

tyxla commented on client/me/reauth-required/security-key-form.scss:
```diff
@@ -0,0 +1,21 @@
+	width: 100%;
+}
+.security-key-form__verification-code-form p {
+	margin-bottom: 0.5em !important;
```
Any good reason to have the `!important` here? Perhaps if we use a classname it won't be necessary?

xknown replied:
I added it because the styles weren't being applied. Does it means we should use a class name for each `p` tag?

xknown replied:
Removed this in 003d81f92d

xknown replied:
Fixed this in 003d81f92d.

tyxla replied:
Thanks - yeah, there is usually a better way to make it work than with `!important` - usually by adding a more specific selector.

tyxla commented on client/me/reauth-required/two-factor-actions.jsx:
```diff
@@ -0,0 +1,99 @@
+
+import { Button, Card } from '@automattic/components';
+import { localize } from 'i18n-calypso';
+import { recordTracksEventWithClientId as recordTracksEvent } from 'state/analytics/actions';
```
Since there is a `recordTracksEvent`, we should be careful with aliasing something else to that name. We could just reference this as `this.props. recordTracksEventWithClientId` in the component and that would do the job.

xknown replied:
Adressed in d96147e3b119919a47c3ab4b09f5840b0b58fa20

tyxla commented on client/me/reauth-required/two-factor-actions.jsx:
```diff
@@ -0,0 +1,99 @@
+/**
+ * External dependencies
+ */
+
```
```suggestion
```

tyxla reviewed (commented):
Thanks a bunch for working on this @xknown 🙌 

For some reason, I'm unable to test this locally or on calypso.live because the two-factor auth cookie never gets stored, and I get the popup every time. I've noticed that this isn't related with this PR, and I can reproduce it on `master` too. I think this is the same thing that @yuliyan reported earlier in Slack - perhaps it's because of Chrome's new SameSite policies for cookies.

Perhaps we can get someone else to test it?

In the meantime, I've left some feedback on the code. Let me know if I can aid with anything.

xknown committed:
Remove leftover code.

xknown commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +68,69 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+		.then( assertion => {
+			const response = assertion.response;
+			if ( typeof response.userHandle !== 'undefined' && null === response.userHandle ) {
+				delete response.userHandle;
```
It looks like this is a leftover of a previous attempt that was causing signature verification issues on the server side. This was removed in 381b813292737fb0be21afca5ebe1ec76498eb82

xknown commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +68,69 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+		.then( response => {
+			const parameters = response.data || [];
+			this.data.two_step_webauthn_nonce = parameters.two_step_nonce;
+			if ( typeof this.data.two_step_webauthn_nonce === 'undefined' ) {
```
It's probably fine if we leave it as is. Even if we don't change `this.data.two_step_webauthn_nonce` on an invalid response, the server side code will only accept these tokens only once on the subsequent requests.
I can move this statement below the `if` block if you think that's more clear.

tyxla replied:
Not a big deal, I think we can leave it as-is 😉

xknown committed:
Use optional chaining and null coalesce

Co-Authored-By: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Remove empty line

Co-Authored-By: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Do not define postLoginRequest for each function call.

xknown commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +68,69 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
 };
 
+TwoStepAuthorization.prototype.loginUserWithSecurityKey = function( args ) {
+	const postLoginRequest = function( endpoint, data ) {
```
You are right. Fixed in d2a9f87e332d1357a37c1827accc0b1b667456fd

xknown committed:
Fix Tracks events

This remove the unneeded aliasing on the import statement. It also logs
different events depending on the button clicked.

The event names are hardcoded so it can be easily found if required.

xknown commented on client/me/reauth-required/two-factor-actions.jsx:
```diff
@@ -0,0 +1,99 @@
+	recordButtonClicked = event => {
+		event.preventDefault();
+
+		this.props.recordTracksEvent( 'calypso_login_two_factor_switch_to_sms_link_click' );
```
Fixed in d96147e3b119919a47c3ab4b09f5840b0b58fa20

xknown commented on client/me/reauth-required/two-factor-actions.jsx:
```diff
@@ -0,0 +1,99 @@
+					<Button
+						data-e2e-link="2fa-security-key-link"
+						value="webauthn"
+						onClick={ this.recordButtonClicked }
```
Addressed in d96147e3b119919a47c3ab4b09f5840b0b58fa20

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -239,21 +223,70 @@ const ReauthRequired = createReactClass( {
+				<TwoFactorActions
+					twoFactorAuthType={ this.state.twoFactorAuthType }
+					onChange={ this.handleAuthSwitch }
+					isSmsSupported={ method === 'sms' || method === 'authenticator' }
```
This is to avoid breaking the current behavior. That is, that even if your account is set up using an authenticator application, you can still request the codes to be sent by SMS. There's some more details in #36809

xknown committed:
Reorder imports.

xknown committed:
Refactor postLoginRequest

This moves the invalid nonce check above so we don't need to declare
variables that we aren't going to use.

xknown commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +68,69 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+			two_step_nonce: this.getTwoStepWebauthnNonce(),
+			...data,
+		};
+		if ( ! _data.two_step_nonce ) {
```
Fixed in 308cd62d6544c8a00341c6a32128aa9ecae02dc3

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -135,25 +146,6 @@ const ReauthRequired = createReactClass( {
 		return this.state.code.length && this.state.code.length > 5;
 	},
 
-	renderSendSMSButton: function() {
```
We are not really removing existing functionality. This is replaced by the new `Send code via text message` button.

tyxla replied:
Gotcha, makes sense.

xknown committed:
Rename tracks event to avoid a warning.

It did not match /^calypso(?:_[a-z]+){2,}$/ and is not a listed
exception.

xknown committed:
Refactor the security key form

This removes unused properties and fixes some code style issues.

xknown committed:
Reorder dependencies

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -26,14 +27,17 @@ import observe from 'lib/mixins/data-observe';
  * Style dependencies
  */
 import './style.scss';
+import TwoFactorActions from 'me/reauth-required/two-factor-actions';
```
Fixed in b5b52655b1

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -26,14 +27,17 @@ import observe from 'lib/mixins/data-observe';
 
 // autofocus is used for tracking purposes, not an a11y issue
-/* eslint-disable jsx-a11y/no-autofocus */
+/* eslint-disable jsx-a11y/no-autofocus, react/prefer-es6-class, jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions, jsx-a11y/anchor-is-valid */
```
That's fine with me, however I think the `ci/circleci: lint-and-translate` check will fail. Is that ok for merging this PR once it's accepted?

tyxla replied:
As long as we're not introducing any of the errors/warnings that it's failing for, we are good. Ideally we should be fixing them as we go, but it would make this PR huge, so I'm fine with ignoring them here.

xknown replied:
The lint warnings are all related to existing code (mostly [this](https://github.com/Automattic/wp-calypso/blob/1835f94661068d5d10323fd42608596bb9064a01/client/me/reauth-required/index.jsx#L202)). I agree we should be fixing them, but IMO that should be part of another PR.

tyxla replied:
Agreed 👍

xknown committed:
Remove unused variable

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -44,6 +48,8 @@ const ReauthRequired = createReactClass( {
 			smsRequestsAllowed: true, // Can the user request another SMS code?
 			smsCodeSent: false,
+			twoFactorAuthType: 'authenticator',
+			webauthnError: false,
```
No, removed in 0a07c3c3fc. Thanks!

xknown committed:
Remove unused code and fix code style issues

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -239,21 +223,70 @@ const ReauthRequired = createReactClass( {
+	renderSecurityKey() {
+		return (
+			<SecurityKeyForm
+				loginUserWithSecurityKey={ () => {
```
Fixed in 79b133377e

xknown committed:
Fix some code style issues

xknown commented on client/me/reauth-required/index.jsx:
```diff
@@ -239,21 +223,70 @@ const ReauthRequired = createReactClass( {
+						this.state.smsRequestsAllowed ||
+						( method === 'sms' && this.state.twoFactorAuthType === 'webauthn' )
+					}
+					isSecurityKeySupported={ isSecurityKeySupported }
```
Update the code in 0874c52dd8. Added also a comment why we require to do this.

xknown commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +68,69 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+			formData.set( key, _data[ key ] );
+		}
+		// eslint-disable-next-line no-undef
+		return fetch( url, {
```
It can be probably migrated to the library you mention, but IMHO it'd also add some overhead we don't really require here.

tyxla replied:
Fair enough 👍

xknown committed:
Fix some code style issues

xknown commented:
> Thanks a bunch for working on this @xknown 🙌
> 
> For some reason, I'm unable to test this locally or on calypso.live because the two-factor auth cookie never gets stored, and I get the popup every time. I've noticed that this isn't related with this PR, and I can reproduce it on `master` too. I think this is the same thing that @yuliyan reported earlier in Slack - perhaps it's because of Chrome's new SameSite policies for cookies.
> 
> Perhaps we can get someone else to test it?
> 
> In the meantime, I've left some feedback on the code. Let me know if I can aid with anything.

Hi @tyxla, many thanks for the detailed review :). I addressed most of the provided feedback. When you have some free time, can you please take another look?

xknown commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +68,69 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+	return postLoginRequest( 'webauthn-challenge-endpoint', {} )
+		.then( response => {
+			const parameters = response.data || [];
+			this.data.two_step_webauthn_nonce = parameters.two_step_nonce;
```
I guess it can be undefined if `TwoStepAuthorization.prototype.fetch` fails. However, the whole `re-auth` dialog isn't shown/rendered if that's the case (which is a bug somewhere else).

xknown committed:
Use the local variable instead.

xknown committed:
Remove the styles for the `p` tag.

blowery commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +67,70 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+
+	const url = 'https://wordpress.com/wp-login.php?action=' + endpoint;
+	// eslint-disable-next-line no-undef
+	const formData = new FormData();
```
this should probably be `window.FormData`, which would let us drop the eslint disable on the previous line.

xknown replied:
Addressed in 89714c680eec7ca23b470420219f92403d931bb6

blowery commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +67,70 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+		client_secret: config( 'wpcom_signup_key' ),
+		auth_type: 'webauthn',
+		two_step_nonce: this.getTwoStepWebauthnNonce(),
+		...data,
```
putting the ellipsis here allows `data` to overwrite the settings from above. Is that what you intended? 

```
$ node
> var foo = { one: 1 }
undefined
> foo
{ one: 1 }
> var bar = { ...foo, one: 2 }
undefined
> bar
{ one: 2 }
> var baz = { one: 3, ...foo }
undefined
> baz
{ one: 1 }
```

xknown replied:
It was intended in a previous iteration, but it's not required anymore. Fixed in 53bed2a267f3bf18ea1721d84f8227640c5877eb.

blowery commented on client/lib/two-step-authorization/index.js:
```diff
@@ -234,6 +300,14 @@ TwoStepAuthorization.prototype.isTwoStepSMSEnabled = function() {
 };
 
+TwoStepAuthorization.prototype.isSecurityKeyEnabled = function() {
+	return this.data?.two_step_webauthn_enabled ?? false;
```
seeing the `data?` here made me realize that `data` can be undefined / null and made me question all the direct setting above that @tyxla also pointed out, which made me have to go through every flow and revalidate it. 

Can we alter the code so that `this.data` is always defined? That would save future folks from having to do all the mental gymnastics to validate that the code is correct.

xknown replied:
Thanks, this was fixed in 89714c6

blowery commented on client/me/reauth-required/index.jsx:
```diff
@@ -86,6 +92,11 @@ const ReauthRequired = createReactClass( {
 				}
 			);
 		}
+		if ( this.state.twoFactorAuthType === 'sms' ) {
```
Maybe we should revisit that convention. I don't often see it in Calypso, especially with triple equals.

tyxla replied:
👍 I'd be really happy to see Yoda no more.

xknown committed:
Make sure this.data is always defined in TwoStepAuthorization

xknown committed:
Move the ellipsis up, overriding items is not needed anymore.

xknown commented:
Thanks for the review @blowery! I addressed all the feedback you provided.

xknown committed:
Reorder the propertyes on the 2fa actions

tyxla commented on client/lib/two-step-authorization/index.js:
```diff
@@ -65,6 +67,71 @@ TwoStepAuthorization.prototype.fetch = function( callback ) {
+			if ( typeof response.success === 'undefined' || ! response.success ) {
+				return Promise.reject( response );
+			}
+			if ( this.isReauthRequired() ) {
```
Would be nice if we reuse this piece, it's used in `validateCode` too.

xknown replied:
Addressed in f7df763efa012a7968abf2947c870524d57f9ad0

tyxla commented on client/me/reauth-required/index.jsx:
```diff
@@ -239,21 +228,61 @@ const ReauthRequired = createReactClass( {
+		);
+	},
+
+	renderSecurityKey() {
```
It almost seems like this is unnecessarily in its own method as it's too short and just feels like unnecessary indirection.

xknown replied:
Removed this method in 2867a000fa9f136aa092c5d0a864937afca51ab1

tyxla reviewed (commented):
I don't have anything big to suggest here, I think the code is looking good, nice work @xknown!

As long as someone tests this and it works well for them, I think we should be good to go!

xknown committed:
Remove the `renderSecurityKey` method.

xknown committed:
Avoid duplicate code

xknown committed:
Reorder imports

xknown committed:
Refresh the webauthn nonce if for some reason it's expired.

There are some cases in which the nonce is invalidated, which makes the
security key form fail until one manually forces a hard refresh. This
change attempts to automatically request a new nonce when the backend
rejects the request mentioning an invalid nonce.

xknown committed:
Replace `findIndex` by `some`, props @blowery

xknown commented:
> `.some` would be a touch faster here, it can abort iterating when it finds the first occurrence of `invalid_two_step_nonce`.

Thanks. Replaced it in 29649b21de

blowery commented:
To echo @tyxla, this looks good, but I don't have a hardware key to test with. If it works for you, feel free to ship it.

xyu commented:
Just tested in 🔥🦊 and works with my YubiKey 5.

xknown commented:
Thank you all for the review   🙌, going to merge it now!
